### ISSUE TITLE
[Question] Extension of the functions by (all) supported API methods/properties

### ISSUE THREAD

**Issue by eizedev** [dnewsholme/PasswordState-Management#96]
Hi @dnewsholme 
First of all, this is a great module, thanks for that!
Your module was so good during the first tests that we internally decided to use it for our environment.
Since a few days I am internally extending the module with, as far as possible and useful, all currently integrated and supported functions of the API and WinAPI.
I will stick to your existing specifications and guidelines for all changes (or try to ;-)).

Of course we could also use the `*-PasswordStateResource` functions to make custom queries against the API but we had internally decided to package the calls into single functions. This makes it easier for everyone to query the API.

I would like to add the changes via a pull request step by step to your module. Before I open a Pull Request, I just wanted to ask you if it is ok for you.  
Therefore I have already provided some changes in a fork (So you can get a feeling for what I mean by that). If it's okay for you, I'd like you to take a look at it. --> https://github.com/eizedev/PasswordState-Management

**[What was changed](https://github.com/dnewsholme/PasswordState-Management/compare/master...eizedev:master):**

- Added new function for creating password dependencies
  - [New-PasswordStateDependency.ps1](https://github.com/eizedev/PasswordState-Management/blob/master/functions/New-PasswordStateDependency.ps1)
  - Including doc: [New-PasswordStateDependency.md](https://github.com/eizedev/PasswordState-Management/blob/master/docs/New-PasswordStateDependency.md)
- Changed [New-PasswordStateFolder.ps1](https://github.com/eizedev/PasswordState-Management/blob/master/functions/New-PasswordStateFolder.ps1)
  - Fixed:
    - `CopySettingsFromPasswordListID` does not exist for folders, removing it (it only exists on password lists).
  - Added:
    - Added the following new possible fields/options for creating folders through the api:
    ```
    CopyPermissionsFromPasswordListID
    CopyPermissionsFromTemplateID
    Guide
    PropagatePermissions
    ```
    - Also added API Key to body if authenticated only with the api key (this is best practise of ClickStudios)
  - Including doc: [New-PasswordStateFolder.md](https://github.com/eizedev/PasswordState-Management/blob/master/docs/New-PasswordStateFolder.md)
- Updated [New-PasswordStateList.ps1](https://github.com/eizedev/PasswordState-Management/blob/master/functions/New-PasswordStateList.ps1)
  - Support all documented and some (currently not working) undocumented properties
  ```powershell
  CopySettingsFromPasswordListID
  CopyPermissionsFromPasswordListID 
  CopySettingsFromTemplateID 
  CopyPermissionsFromTemplateID
  LinkToTemplate
  PrivatePasswordList
  ApplyPermissionsForUserID 
  ApplyPermissionsForSecurityGroupID 
  ApplyPermissionsForSecurityGroupName 
  Permission
  ImageFileName
  SiteID 
  Guide
  ```
  - For the "currently not working or undocumented" properties i have opened a ticket at ClickStudios
  - Including doc: [New-PasswordStateList.md](https://github.com/eizedev/PasswordState-Management/blob/master/docs/New-PasswordStateList.md)
- Added possibility to sort the response/returned object by applying `-Sort`
  - This is also integrated to sort the `body` for verbose output
  - For this, added a function [Get-PSCustomObject](https://github.com/eizedev/PasswordState-Management/blob/master/internal/functions/Get-PSCustomObject.ps1) to sort existing PSCustomObjects
    - `$output | Get-PSCustomObject -Sort`
    - This is by default integrated in [New-PasswordStateResource.ps1 and Get-PasswordStateResource.ps1](https://github.com/eizedev/PasswordState-Management/commit/cfe1240986dc590bb6e40f4c284566c65d6c7006#diff-6c580cf902a9e40eb7eb1015127134f0)
- Updated all Markdown Documentation files under `docs`, added doc for already existing functions,  removed documentation for no longer available functions
- Added/Updated the [external help](https://github.com/eizedev/PasswordState-Management/tree/master/en-us) file using PlatyPS `New-ExternalHelp` in the `en-US` folder
- Fixed `PreventAuditing` on [Get-PasswordStateList](https://github.com/eizedev/PasswordState-Management/commit/bb45686ba3ca3a8d55bb9e95f1e1df1eefba4939#diff-f43b978c186a666f368759c03f231e88) if only PasswordListID was specified (PreventAuditing is not possible then)
- Added [aliases](https://github.com/dnewsholme/PasswordState-Management/commit/61e17bcba780371eab0f31d7ab0242076ecc41e3) `Find-PasswordstateList` and `Find-PasswordStateFolder`

I haven't integrated any pester tests yet, I hadn't worked with it that much, but I want to have a look at it or maybe someone else could do it then?
I think those were all the latest changes. The new/changed functions are also already in use and tested.

What do you say?

EDIT: The newly added `online version` url in the markdown files will be changed to your repository before opening a pull request.

Thanks & Regards,
René

**Comment by trir262** [comment]
@eizedev I'd be happy to write pester tests for the new functions as I am busy (although got a little distracted lately) by writing tests for the already existing ones.

**Comment by eizedev** [comment]
@trir262 Ah cool, thanks!

**Comment by dnewsholme** [comment]
@eizedev Thanks for your contribution. It looks great, I'm happy with everything you've done so far it looks good and keeps to the style of the functions. Send over the pull request when you are ready and ill review.

I think we should change both from Update to Set as the verb but alias back to update to maintain compatability.

@trir262  Thanks, you've already done some great work on the pester tests.

**Comment by eizedev** [comment]
@dnewsholme Thanks :) I have also renamed `Update-PasswordStatePassword` to `Set-PasswordStatePassword`, renamed Exported function, pester test and exclusionlist and added the alias.

Pull request opened #97

**Comment by dnewsholme** [comment]
closed in #97
